interfaces/builtin/utils.go

```diff
@@ -32,46 +32,52 @@ import (
 // The maximum number of Usb bInterfaceNumber.
 const UsbMaxInterfaces = 32
 
-// AppLabelExpr returns the specification of the apparmor label describing
-// all the apps bound to a given slot. The result has one of three forms,
+// labelExpr returns the specification of the apparmor label describing
+// given apps and hooks. The result has one of three forms,
 // depending on how apps are bound to the slot:
 //
 // - "snap.$snap_instance.$app" if there is exactly one app bound
-// - "snap.$snap_instance.{$app1,...$appN}" if there are some, but not all, apps bound
-// - "snap.$snap_instance.*" if all apps are bound to the slot
-func appLabelExpr(apps map[string]*snap.AppInfo, snap *snap.Info) string {
+// - "snap.$snap_instance.{$app1,...$appN, $hook1...$hookN}" if there are some, but not all, apps/hooks bound
+// - "snap.$snap_instance.*" if all apps/hook are bound to the plug or slot
+func labelExpr(apps map[string]*snap.AppInfo, hooks map[string]*snap.HookInfo, snap *snap.Info) string {
 	var buf bytes.Buffer
+
+	names := make([]string, 0, len(apps)+len(hooks))
+	for appName := range apps {
+		names = append(names, appName)
+	}
+	for hookName := range hooks {
+		names = append(names, fmt.Sprintf("hook.%s", hookName))
+	}
+	sort.Strings(names)
+
 	fmt.Fprintf(&buf, `"snap.%s.`, snap.InstanceName())
-	if len(apps) == 1 {
-		for appName := range apps {
-			buf.WriteString(appName)
-		}
-	} else if len(apps) == len(snap.Apps) {
+	if len(names) == 1 {
+		buf.WriteString(names[0])
+	} else if len(apps) == len(snap.Apps) && len(hooks) == len(snap.Hooks) {
 		buf.WriteByte('*')
-	} else {
-		appNames := make([]string, 0, len(apps))
-		for appName := range apps {
-			appNames = append(appNames, appName)
-		}
-		sort.Strings(appNames)
+	} else if len(names) > 0 {
```

Comment:
Note, the `... len(names)>0` here is guarding against the bug fixed by #7836; however, I'm inclined to leave it to guard against generating an invalid `snap.foo.{}` label (which would be rejected by apparmor and fail load-profiles) in case names is ever empty for whatever reason. With the guard in place we'll get `snap.foo.` that's not going to match anything in the system. cc @zyga